src/core/shape_inference/include/interpolate_shape_inference.hpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <openvino/op/interpolate.hpp>
+
+#include "utils.hpp"
+
+namespace ov {
+namespace op {
+
+namespace v0 {
+template <class T>
+void shape_infer(const Interpolate* op,
+                 const std::vector<T>& input_shapes,
+                 std::vector<T>& output_shapes,
+                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data) {
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 2 && output_shapes.size() == 1);
+
+    const auto& input_shape = input_shapes[0];
+    auto& output_shape = output_shapes[0];
+    output_shape = input_shape;
+
+    if (input_shape.rank().is_static()) {
+        auto input_rank = input_shape.size();
+        NODE_VALIDATION_CHECK(op,
+                              std::all_of(op->m_attrs.axes.begin(),
+                                          op->m_attrs.axes.end(),
+                                          [input_rank](size_t axis) {
+                                              return axis < input_rank;
+                                          }),
+                              "Axis value should less than input rank. ",
+                              "Got: input rank ",
+                              input_rank,
+                              ", axes ",
+                              op->m_attrs.axes);
+
+        T target_spatial_shape;
+        if (get_data_as_shape<T>(1, op, target_spatial_shape, constant_data)) {
+            size_t i = 0;
+            for (auto axis : op->m_attrs.axes) {
+                output_shape[axis] = target_spatial_shape[i++];
+            }
+        } else {
+            for (auto axis : op->m_attrs.axes) {
+                output_shape[axis] = ov::Dimension::dynamic();
+            }
+        }
+    }
+}
+}  // namespace v0
+
+namespace v4 {
+template <typename T>
+std::vector<T> correct_pad(const std::vector<T>& p, size_t rank) {
+    size_t pad_len = p.size();
+    if (pad_len == rank) {
+        return p;
+    }
+
+    std::vector<T> result;
+    if (pad_len > rank) {
+        result.insert(result.end(), p.begin(), p.begin() + rank);
+    } else {
+        result = p;
+        result.insert(result.end(), rank - pad_len, T{});
+    }
+    return result;
+}
```

Comment:
Looks like you need this function only for `pads_begin` and `pads_end` modifying, which are of type `std::vector<size_t>`. So I suggest not using templates here.
Also, it's probably good idea to modify `p` in this function to avoid unnecessary copying

Reply:
Yes, you are right, I removed this function and move the logic to correct_pads_attr to make code more readable